File: sdk/typescript/src/tracer.ts
export const enum SpanKind {
  INTERNAL = 1,
  SERVER = 2,
  CLIENT = 3,
  PRODUCER = 4,
  CONSUMER = 5,
}

export const enum StatusCode {
  UNSET = 0,
  OK = 1,
  ERROR = 2,
}

export interface SpanOptions {
  kind?: SpanKind;
  attributes?: Record<string, string | number | boolean>;
}

export interface SpanEvent {
  name: string;
  timestamp: number;
  attributes: Record<string, string>;
}

export class Span {
  readonly traceId: string;
  readonly spanId: string;
  readonly name: string;
  readonly kind: SpanKind;
  readonly startTimeNs: bigint;
  endTimeNs: bigint = 0n;
  statusCode: StatusCode = StatusCode.UNSET;
  statusMessage = '';
  readonly attributes: Map<string, string | number | boolean> = new Map();
  readonly events: SpanEvent[] = [];
  private ended = false;

  constructor(traceId: string, spanId: string, name: string, kind: SpanKind) {
    this.traceId = traceId;
    this.spanId = spanId;
    this.name = name;
    this.kind = kind;
    this.startTimeNs = BigInt(Date.now()) * 1_000_000n;
  }

  setAttributes(attrs: Record<string, string | number | boolean>): void {
    for (const [k, v] of Object.entries(attrs)) {
      this.attributes.set(k, v);
    }
  }

  addEvent(name: string, attributes: Record<string, string> = {}): void {
    this.events.push({ name, timestamp: Date.now(), attributes });
  }

  recordException(error: Error): void {
    this.addEvent('exception', {
      'exception.type': error.name,
      'exception.message': error.message,
    });
  }

  setStatus(code: StatusCode, message = ''): void {
    this.statusCode = code;
    this.statusMessage = message;
  }

  end(): void {
    if (!this.ended) {
      this.ended = true;
      this.endTimeNs = BigInt(Date.now()) * 1_000_000n;
    }
  }
}

export interface TracerConfig {
  endpoint?: string;
  service?: string;
  samplingRate?: number;
}

export class Tracer {
  private readonly config: Required<TracerConfig>;

  constructor(config: TracerConfig = {}) {
    this.config = {
      endpoint: config.endpoint ?? 'http://localhost:4318',
      service: config.service ?? '',
      samplingRate: config.samplingRate ?? 1.0,
    };
  }

  startSpan(name: string, options: SpanOptions = {}): Span {
    const traceId = generateId(32);
    const spanId = generateId(16);
    const span = new Span(traceId, spanId, name, options.kind ?? SpanKind.INTERNAL);
    if (options.attributes) {
      span.setAttributes(options.attributes);
    }
    return span;
  }
}

function generateId(length: number): string {
  const chars = '0123456789abcdef';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += chars[Math.floor(Math.random() * 16)];
  }
  return result;
}

File: sdk/python/ginger/tracer.py
"""Tracer implementation for the Ginger Python SDK."""

import os
import time
import uuid
from enum import IntEnum
from typing import Any


class SpanKind(IntEnum):
    INTERNAL = 1
    SERVER = 2
    CLIENT = 3
    PRODUCER = 4
    CONSUMER = 5


class StatusCode(IntEnum):
    UNSET = 0
    OK = 1
    ERROR = 2


class Span:
    """Represents a single traced operation."""

    def __init__(
        self,
        trace_id: str,
        span_id: str,
        name: str,
        kind: SpanKind = SpanKind.INTERNAL,
    ) -> None:
        self.trace_id = trace_id
        self.span_id = span_id
        self.name = name
        self.kind = kind
        self.start_time_ns = time.time_ns()
        self.end_time_ns: int = 0
        self.status_code = StatusCode.UNSET
        self.status_message = ""
        self.attributes: dict[str, Any] = {}
        self.events: list[dict[str, Any]] = []
        self._ended = False

    def set_attribute(self, key: str, value: Any) -> None:
        self.attributes[key] = value

    def set_status(self, code: StatusCode, message: str = "") -> None:
        self.status_code = code
        self.status_message = message

    def add_event(self, name: str, attributes: dict[str, str] | None = None) -> None:
        self.events.append(
            {"name": name, "timestamp": time.time_ns(), "attributes": attributes or {}}
        )

    def record_exception(self, exc: BaseException) -> None:
        self.add_event(
            "exception",
            {"exception.type": type(exc).__name__, "exception.message": str(exc)},
        )

    def end(self) -> None:
        if not self._ended:
            self._ended = True
            self.end_time_ns = time.time_ns()

    def __enter__(self) -> "Span":
        return self

    def __exit__(self, *args: Any) -> None:
        self.end()


class Tracer:
    """Creates and manages spans."""

    def __init__(
        self,
        endpoint: str = "http://localhost:4318",
        service: str = "",
        sampling_rate: float = 1.0,
    ) -> None:
        self.endpoint = endpoint
        self.service = service or os.environ.get("OTEL_SERVICE_NAME", "unknown")
        self.sampling_rate = sampling_rate

    def start_span(
        self,
        name: str,
        kind: SpanKind = SpanKind.INTERNAL,
        attributes: dict[str, Any] | None = None,
    ) -> Span:
        trace_id = uuid.uuid4().hex
        span_id = uuid.uuid4().hex[:16]
        span = Span(trace_id, span_id, name, kind)
        if attributes:
            for k, v in attributes.items():
                span.set_attribute(k, v)
        return span
